cpp/src/tsvd/tsvd_mg.cu

```diff
@@ -0,0 +1,436 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <common/cumlHandle.hpp>
+#include <common/cuml_comms_int.hpp>
+#include <common/device_buffer.hpp>
+#include <cuda_utils.cuh>
+#include <cuml/common/cuml_allocator.hpp>
+#include <cuml/decomposition/sign_flip_mg.hpp>
+#include <cuml/decomposition/tsvd.hpp>
+#include <cuml/decomposition/tsvd_mg.hpp>
+#include <linalg/eltwise.cuh>
+#include <matrix/math.cuh>
+#include <opg/linalg/mm_aTa.hpp>
+#include <opg/stats/mean.hpp>
+#include <opg/stats/mean_center.hpp>
+#include <opg/stats/stddev.hpp>
+#include <stats/mean_center.cuh>
+#include "tsvd.cuh"
+
+using namespace MLCommon;
+
+namespace ML {
+namespace TSVD {
+namespace opg {
+
+template <typename T>
+void fit_impl(cumlHandle &handle, std::vector<Matrix::Data<T> *> &input_data,
+              Matrix::PartDescriptor &input_desc, T *components,
+              T *singular_vals, paramsTSVD prms, cudaStream_t *streams,
+              int n_streams, bool verbose) {
+  const MLCommon::cumlCommunicator &comm = handle.getImpl().getCommunicator();
+  cublasHandle_t cublas_handle = handle.getImpl().getCublasHandle();
+  const std::shared_ptr<deviceAllocator> allocator =
+    handle.getImpl().getDeviceAllocator();
+
+  int len = prms.n_cols * prms.n_cols;
```

Comment:
The use of `int` here is precisely the type of case that will overflow quickly as `n_cols` grows. Linking in https://github.com/rapidsai/cuml/issues/2459

Reply:
Do we want to do that in this PR?

Reply:
Doesn’t need to be done in this pr. I wanted to reference the issue here for future tracking.